fe/estimator_abfe.py

```diff
@@ -300,51 +240,57 @@ def _deltaG(model, sys_params) -> Tuple[Tuple[float, List], np.array]:
         for future in futures:
             results.append(future.result())
 
-    mean_du_dls = []
     all_grads = []
 
     if model.endpoint_correct:
-        ti_results = results[:-1]
+        sim_results = results[:-1]
     else:
-        ti_results = results
+        sim_results = results
 
-    for lambda_idx, (lambda_window, result) in enumerate(zip(model.lambda_schedule, ti_results)):
-        # (ytz): figure out what to do with stddev(du_dl) later
-        # print(f"{model.prefix} index {lambda_idx} lambda {lambda_window:.5f} <du/dl> {np.mean(result.du_dls):.5f} med(du/dl) {np.median(result.du_dls):.5f}  o(du/dl) {np.std(result.du_dls):.5f}")
-        mean_du_dls.append(np.mean(result.du_dls))
+    U_kn = []
+    N_ks = []
+    for lambda_idx, (lambda_window, result) in enumerate(zip(model.lambda_schedule, sim_results)):
+        U_kn.append(result.lambda_us)
+        N_ks.append(len(result.lambda_us))
         all_grads.append(result.du_dps)
 
-    tibar_dG = 0
+    K = len(model.lambda_schedule)
+    U_kn = np.array(U_kn) # [K, F, K]
+    U_kn = np.reshape(U_kn, (-1, K)).transpose() # [K, F*K]
+    # deal with overflow issues here, tbd: make it more robust?
+    U_kn = np.where(U_kn < -1e10, np.inf, U_kn)
+
+    mbar = pymbar.MBAR(U_kn, N_ks)
+    differences, error_estimates = mbar.getFreeEnergyDifferences()
+    f_k, error_k = differences[0], error_estimates[0]
+    mbar_dG = f_k[-1]
+    mbar_dG_err = error_k[-1]
+
+
     bar_dG = 0
     bar_dG_err = 0
-
     for lambda_idx in range(len(model.lambda_schedule) - 1):
-        # tibar
-        lamb_start = model.lambda_schedule[lambda_idx]
-        lamb_end = model.lambda_schedule[lambda_idx+1]
-        delta_lamb = lamb_end - lamb_start
-        fwd_work_approx = ti_results[lambda_idx].du_dls*delta_lamb
-        rev_work_approx = -ti_results[lambda_idx+1].du_dls*delta_lamb
-        tibar, ti_bar_err = pymbar.BAR(model.beta*fwd_work_approx, model.beta*rev_work_approx)
-        tibar_overlap = endpoint_correction.overlap_from_cdf(fwd_work_approx, rev_work_approx)
-        tibar_dG += tibar/model.beta
-
-        # exact_bar
-        fwd_work_exact = model.beta*ti_results[lambda_idx].right_dus
-        rev_work_exact = model.beta*ti_results[lambda_idx+1].left_dus
-
-        dG_exact, exact_bar_err = pymbar.BAR(fwd_work_exact, rev_work_exact)
+        # snapshots from lamb, evaluated at U(lamb+1) - U(lamb)
+        fwd_delta_u = model.beta*(sim_results[lambda_idx].lambda_us[lambda_idx+1] - sim_results[lambda_idx].lambda_us[lambda_idx])
+        # snapshots from lamb+1, evaluated at U(lamb) - U(lamb+1)
+        rev_delta_u = model.beta*(sim_results[lambda_idx+1].lambda_us[lambda_idx] - sim_results[lambda_idx+1].lambda_us[lambda_idx+1])
```

Comment:
Looks correct after some staring, but to reduce the number of places per line where sign errors or off-by-one indexing errors could possibly hide, we may want to extract smaller steps, perhaps as
```python
    def delta_U(from_idx, to_idx):
        """[U(x, to_idx) - U(x, from_idx) for x in xs]
        where xs are simulated at from_idx"""
        current = sim_results[from_idx]
        current_energies = current.lambda_us[from_idx]
        perturbed_energies = current.lambda_us[to_idx]
        return perturbed_energies - current_energies
    
    bar_dG = 0
    bar_dG_err = 0
    for lambda_idx in range(len(model.lambda_schedule) - 1):
        fwd_delta_u = model.beta * delta_U(lambda_idx, lambda_idx + 1)
        rev_delta_u = model.beta * delta_U(lambda_idx + 1, lambda_idx)
```

Reply:
I'll clean this up!